### ISSUE TITLE
Pass sample weights to custom objective functions in the python sklearn interface

### ISSUE THREAD

**Issue by bemao** [dmlc/xgboost#5768]
Currently, the python sklearn API for XGBRegessor works by passing the expected arguments for the sklearn-formatted objective function `(ylabel, ypred)` via the decorator `_objective_decorator(func).` This function does not allow for passing sample weights (`dmatrix.get_weight()`) to the custom objective function.

I suggest changing the `_objective_decorator` to allow for custom objective functions that make use of sample weights.

**Comment by trivialfis** [comment]
That way all objective defined in scikit learn will break ... Maybe we need to have 2 different decorators.

**Comment by bemao** [comment]
I was thinking the the decorator would check whether or not the user-defined objective accepts a `weights` argument and if so, passes the `sample_weight` to it, otherwise it will work as it is currently defined.

**Comment by pidefrem** [comment]
As far as I understand, applying `sample_weight` only consist of multiplying the gradient and the hessian by the sample_weight (this is how it is done on c++ side). So we can just modify the decorator as suggested.

Here is my proposal:
```
def _objective_decorator(func):
    def inner(preds, dmatrix):
        """internal function"""
        labels = dmatrix.get_label()
        grad, hess = func(labels, preds)
        
        weights = dmatrix.get_weight()
        if weights.shape[0] > 0:
          if weights.shape[0] != labels.shape[0]:
            raise Exception("Number of weights should be equal to number of data points.")
          grad *= weights
          hess *= weights
        return grad, hess
    return inner
```
Thus we don't have to modify sklearn API.

**Comment by pidefrem** [comment]
As far as I understand, applying `sample_weight` only consist sof multiplying the gradient and the hessian by the sample_weight (this is how it is done on c++ side). So we can just modify the decorator as suggested.

Here is my proposal:
```
def _objective_decorator(func):
    def inner(preds, dmatrix):
        """internal function"""
        labels = dmatrix.get_label()
        grad, hess = func(labels, preds)
        
        weights = dmatrix.get_weight()
        if weights.shape[0] > 0:
          if weights.shape[0] != labels.shape[0]:
            raise Exception("Number of weights should be equal to number of data points.")
          grad *= weights
          hess *= weights
        return grad, hess
    return inner
```
Thus we don't have to modify sklearn API.

**Comment by trivialfis** [comment]
That's not necessarily true for all objectives.  Also there are other meta information like `group` for ranking, `label_upper/lower_bound` for survival.  I think for these cases, using plain `train` function instead of scikit-learn wrapper might be a better way to go, at least for now.
